planner/core/rule_join_elimination.go

```diff
@@ -0,0 +1,187 @@
+		return outerPlan
+	}
+	return p
+}
+
+func (o *outerJoinEliminator) extractInnerJoinKeys(join *LogicalJoin, innerChildIdx int) *expression.Schema {
+	var joinKeys []*expression.Column
+	for _, eqCond := range join.EqualConditions {
+		joinKeys = append(joinKeys, eqCond.GetArgs()[innerChildIdx].(*expression.Column))
+	}
+	return expression.NewSchema(joinKeys...)
+}
+
+func (o *outerJoinEliminator) isAggColsAllFromOuterTable(outerPlan LogicalPlan, aggCols []*expression.Column) bool {
+	if len(aggCols) == 0 {
+		return false
+	}
+	for _, col := range aggCols {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := outerPlan.Schema().FindColumn(columnName); c == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (o *outerJoinEliminator) isParentColsAllFromOuterTable(outerPlan LogicalPlan, parentSchema *expression.Schema) bool {
+	if parentSchema == nil {
+		return false
+	}
+	for _, col := range parentSchema.Columns {
+		columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+		if c, _ := outerPlan.Schema().FindColumn(columnName); c == nil {
+			return false
+		}
+	}
+	return true
+}
+
+func (o *outerJoinEliminator) isInnerJoinKeysContainUniqueKey(innerPlan LogicalPlan, joinKeys *expression.Schema) bool {
+	for _, keyInfo := range innerPlan.Schema().Keys {
+		joinKeysContainKeyInfo := true
+		for _, col := range keyInfo {
+			columnName := &ast.ColumnName{Schema: col.DBName, Table: col.TblName, Name: col.ColName}
+			if c, _ := joinKeys.FindColumn(columnName); c == nil {
+				joinKeysContainKeyInfo = false
+				break
+			}
+		}
+		if joinKeysContainKeyInfo {
+			return true
+		}
+	}
+	return false
+}
+
+func (o *outerJoinEliminator) isInnerJoinKeysContainIndex(innerPlan LogicalPlan, joinKeys *expression.Schema) bool {
+	ds, ok := innerPlan.(*DataSource)
+	if !ok {
+		return false
+	}
+	for _, path := range ds.possibleAccessPaths {
+		if path.isTablePath {
+			continue
+		}
+		idx := path.index
+		if !idx.Unique {
+			continue
+		}
+		joinKeysContainIndex := true
+		for _, idxCol := range idx.Columns {
+			columnName := &ast.ColumnName{Schema: ds.DBName, Table: ds.tableInfo.Name, Name: idxCol.Name}
+			if c, _ := joinKeys.FindColumn(columnName); c == nil {
+				joinKeysContainIndex = false
+				break
+			}
+		}
+		if joinKeysContainIndex {
+			return true
+		}
+	}
+	return false
+}
+
+// Check whether a LogicalPlan is a LogicalAggregation and its all aggregate functions is duplicate agnostic.
+// Also, check all the args are expression.Column.
+func (o *outerJoinEliminator) isDuplicateAgnosticAgg(p LogicalPlan) (_ bool, cols []*expression.Column) {
+	if agg, ok := p.(*LogicalAggregation); ok {
+		cols = agg.groupByCols
+		for _, aggDesc := range agg.AggFuncs {
+			if !aggDesc.HasDistinct &&
+				aggDesc.Name != ast.AggFuncFirstRow &&
+				aggDesc.Name != ast.AggFuncMax &&
+				aggDesc.Name != ast.AggFuncMin {
```

Comment:
Please add test cases to cover all the agg functions with Column/ non-Column args.

And add test case for `count` with a different number of args.